Source/JavaScriptCore/runtime/WaiterListManager.cpp

```diff
@@ -0,0 +1,252 @@
+#include "JSLock.h"
+#include "ObjectConstructor.h"
+#include <wtf/DataLog.h>
+#include <wtf/NeverDestroyed.h>
+#include <wtf/RawPointer.h>
+
+namespace JSC {
+
+namespace WaiterListsManagerInternal {
+static constexpr bool verbose = false;
+}
+
+WaiterListManager& WaiterListManager::singleton()
+{
+    static LazyNeverDestroyed<WaiterListManager> manager;
+    static std::once_flag onceKey;
+    std::call_once(onceKey, [&] {
+        manager.construct();
+    });
+    return manager;
+}
+
+JSValue WaiterListManager::wait(JSGlobalObject* globalObject, VM& vm, void* ptr, AtomicsWaitValidation validation, Seconds timeout, AtomicsWaitType type)
+{
+    if (type == AtomicsWaitType::Async) {
+        JSObject* object = constructEmptyObject(globalObject->vm(), globalObject->objectStructureForObjectConstructor());
+
+        bool isAsync = false;
+        JSValue value;
+        if (validation == AtomicsWaitValidation::Fail)
+            value = vm.smallStrings.notEqualString();
+        else if (!timeout)
+            value = vm.smallStrings.timedOutString();
+        else {
+            isAsync = true;
+            JSPromise* promise = JSPromise::create(vm, globalObject->promiseStructure());
+            addAsyncWaiter(ptr, promise, timeout);
+            value = promise;
+        }
+
+        object->putDirect(vm, vm.propertyNames->async, jsBoolean(isAsync));
+        object->putDirect(vm, vm.propertyNames->value, value);
+        return object;
+    }
+
+    if (validation == AtomicsWaitValidation::Fail)
+        return vm.smallStrings.notEqualString();
+
+    Waiter waiter { &vm };
+    WaiterList* list = findOrCreateList(ptr);
+    const TimeWithDynamicClockType& time = MonotonicTime::now() + timeout;
+
+    {
+        Locker locker { list->lock };
+        list->addLast(locker, &waiter);
+        dataLogLnIf(WaiterListsManagerInternal::verbose, "WaiterListManager added a new SyncWaiter ", RawPointer(&waiter), " to a waiterList for ptr ", RawPointer(ptr));
+
+        while (waiter.getVM() && time.nowWithSameClock() < time)
+            waiter.getCondition().waitUntil(list->lock, time.approximateWallTime());
+        ASSERT(!waiter.getVM() || waiter.getVM() == &vm);
+        bool didGetDequeued = !waiter.getVM();
+
+        if (didGetDequeued)
+            return vm.smallStrings.okString();
+        list->take(locker, &waiter);
+        return vm.smallStrings.timedOutString();
+    }
+}
+
+void WaiterListManager::addAsyncWaiter(void* ptr, JSPromise* promise, Seconds timeout)
+{
+    WaiterList* list = findOrCreateList(ptr);
+    Ref<Waiter> waiter = adoptRef(*new Waiter(promise));
+
+    {
+        Locker locker { list->lock };
+        list->addLast(locker, waiter.ptr());
+
+        if (timeout != Seconds::infinity()) {
+            Ref<RunLoop::DispatchTimer> timer = RunLoop::current().dispatchAfter(timeout, [this, ptr, waiter = waiter.copyRef()] {
+                WaiterList* list = findList(ptr);
+                if (list) {
```

Comment:
I think this is not allowed since at this point there is no guarantee that we are still keeping SharedArrayBuffer.
What happens if this thread is no longer running JSC VM, and just after running `findList`, some other thread destroyed SharedArrayBuffer?
Given that at this point we have no guarantee that we are still having SharedArrayBuffer alive, we need to be super careful about these conditions.

Reply:
I think WaiterList should be `ThreadSafeRefCounted`. And we should keep ref of that when returning it from `findList` etc. So, `findList` returns `RefPtr<WaiterList>`. And `findOrCreateList` returns `Ref<WaiterList>`

Reply:
And this means that we may find that this waiter is not in the list.

1. Until this, SharedArrayBuffer gets destroyed.
2. New SharedArrayBuffer is created and it has the same address
3. So, when getting a list from this pointer here, we get a new list, which does not include this waiter.

We need to handle this kind of case too.